Added python script using solve_bvp validated against Kansal et al., …

amanb1901 opened the pull request:
…2024

Need to delete redundant lines, correct the comments and make nicer plots

amanb1901 committed:
Added python script using solve_bvp validated against Kansal et al., 2024

Need to delete redundant lines, correct the comments and make nicer plots

amanb1901 committed:
Changed solve_bvp settings to solve convergence issues

VatsalSy committed:
Add LE_solver.py implementing Lubrication Equation solver

Implements a simplified solver for the Lubrication Equation (LE) as an
alternative to the Generalized Lubrication Equation (GLE) solver.
Maintains the same GUI/non-GUI options and output structure as GLE_solver.py
but uses the standard lubrication equation formulation.

VatsalSy committed:
Merge branch 'amanb1901-patch-1' of github.com:comphy-lab/Contact-line-subgrid-modeling into amanb1901-patch-1

VatsalSy commented on GLE_solver_v2.py:
```diff
@@ -0,0 +1,170 @@
+# Initial conditions
+h0 = lambda_slip  # h at s = 0
+theta0 = np.pi/2  # theta at s = 0
+w = 0  # curvature boundary condition at s = \Delta, this needs to be not remain constant, but fed back from the DNS
+
+
+
```
we will do it later on

VatsalSy committed:
Remove unit references from GLE_solver_v2.py

- Remove trailing spaces from axis labels
- Remove micrometers (μm) units from axis labels
- Remove 1e6 multipliers for unit conversion
- Update Delta comment to clarify it's dimensionless
- All variables now correctly treated as dimensionless

VatsalSy committed:
Remove unit references from GLE_solver.py

- Remove micrometers (μm) units from axis labels
- Remove 1e6 multipliers for unit conversion in plots
- Update Delta comment to clarify it's dimensionless
- Fix xlim calculations to use dimensionless values
- Maintain consistency with dimensionless approach

VatsalSy committed:
Remove unit references from compare_results.py

- Remove micrometers (μm) units from axis labels
- Remove 1e6 multipliers for unit conversion in plots
- Fix xlim calculations to use dimensionless values
- Ensure consistent dimensionless treatment across Python and C comparisons

VatsalSy committed:
Remove unit reference from test_LE_solver.py

- Remove "4 microns" comment from s_max variable
- Maintain dimensionless approach throughout test files

VatsalSy committed:
Consolidate GLE solver versions

- Merge GLE_solver_v2.py functionality into GLE_solver.py
- Update parameters: Ca=0.0246, lambda_slip=1e-4, theta0=π/2
- Add negative sign to dw_ds equation (receding contact line)
- Use larger integration domain (0 to 10) with 1000 points
- Remove redundant GLE_solver_v2.py file

VatsalSy committed:
Remove LE_solver.py and rename test file

- Remove redundant LE_solver.py
- Rename test_LE_solver.py to validate_LE_solver.py
  (it was not a proper test file but a validation script)

VatsalSy committed:
Add pandas to Python requirements

Required for validate_LE_solver.py to read CSV data

VatsalSy committed:
Update C headers to include omega in CSV output

- Modify gle_io.h to write omega values to CSV
- Update gle_shooting.h to allocate and output omega array
- Ensures C and Python implementations have consistent output format

VatsalSy committed:
Clean up GLE_solver.py formatting and fix integration domain

- Move Delta definition to top with other parameters
- Convert multi-line comments to docstring format
- Fix integration domain to use Delta instead of hardcoded value
- Remove outdated comment about pi/6 initial angle
- Improve code organization and readability

VatsalSy committed:
Fix LaTeX formatting and initial guess order

- Escape backslashes in LaTeX strings for proper rendering
- Fix initial guess array order (was incorrectly swapped)
- Add clarifying comment for f1, f2, f3 functions
- Use LaTeX formatting for lambda_slip and mu_r in plot labels

VatsalSy committed:
Add 2x2 plot grid with new visualizations

- Calculate x(s) = integral of cos(theta) ds for horizontal position
- Convert from 2x1 to 2x2 subplot grid
- Add theta vs h plot to show angle-thickness relationship
- Add h vs x plot to show thickness vs horizontal position
- Update CSV output to include x values
- Improve plot axis limits for better visualization

VatsalSy committed:
Update boundary condition comments in GLE_solver.py

VatsalSy committed:
Update Delta parameter comment in GLE_solver.py for clarity

VatsalSy committed:
Add command-line parameter support to GLE solver

- Add argparse for command-line argument parsing
- Support parameters: --delta, --ca, --lambda-slip, --mu-r, --theta0, --w
- Convert global parameters to defaults that can be overridden
- Update functions to accept parameters instead of using globals
- Display parameters used in output
- Maintain backward compatibility with --gui flag

VatsalSy committed:
Fix BVP solver initial guess variable order

Corrected the initial guess array to match the expected order [h, theta, omega]
used by both the GLE and boundary_conditions functions. The solver is robust
enough to converge even with the wrong order, but this fix ensures consistency.

VatsalSy committed:
Fix exception handling and code structure in validate_LE_solver.py

- Replace bare except clause with specific exceptions (FileNotFoundError, pd.errors.ParserError)
- Remove orphaned docstring and fix indentation issues
- Properly align code blocks for better readability

VatsalSy committed:
Add x₀ vs Ca visualization during continuation method

Track and plot the position of minimum contact angle (x₀) as a function of
Capillary number during parameter continuation. This provides insight into
how the contact line structure evolves as Ca approaches the critical value.

- Add plot_x0_vs_ca() function for creating scatter plots with trend lines
- Modify continuation method to calculate and store x₀ values
- Export x₀ vs Ca data to CSV for further analysis
- Fix minor plotting bug (ax4.set_ylim typo)
- Use log scale for Ca axis when range spans multiple orders of magnitude

VatsalSy committed:
update README

VatsalSy committed:
GLE_solver ready

VatsalSy committed:
first version of continuation method -- already after trying many things

VatsalSy committed:
Clean up and fix test suite

- Remove temporary test files (test_ca_zero*, check_ca0_solution.py)
- Fix all failing tests by updating function signatures
- Add comprehensive test documentation (test/README.md)
- Fix test_x0_definition.py to use new x0 definition
- Update test_GLE_solver.py to pass required parameters (Ca, mu_r, lambda_slip)
- Update test_integration.py to pass all required solver parameters
- Add test_theta_min_continuation.py for continuation method testing
- Include test_x0_definition.png showing validation results

All 43 tests now pass successfully.

VatsalSy committed:
Relocate find_x0_utils.py to src-local for better organization

- Moved find_x0_utils.py from root directory to src-local/
- Created __init__.py to make src-local a Python package
- Updated imports in GLE_solver.py and GLE_continuation_hybrid.py
- This organization keeps utility functions separate from main solvers

VatsalSy committed:
Add debugging for Phase 2 pseudo-arclength continuation

- Added extensive debugging output to trace Phase 2 execution
- Created test_phase2_debug.py to isolate and test the issue
- Updated CLAUDE.md with debugging information and known issues
- Added __all__ exports to src-local/__init__.py
- Fixed import path in test_x0_definition.py
- Enhanced tangent computation with backward difference fallback
- Added progress tracking for predictor-corrector steps

VatsalSy committed:
Fix import paths in test files for src-local module

Updated test files to properly add parent and src-local directories to Python path
- test_phase2_debug.py: Add src-local to path for find_x0_utils import
- test_x0_definition.py: Use absolute path approach instead of relative
- test_GLE_solver.py: Add parent directory to path for imports

VatsalSy committed:
Rename GLE_continuation_hybrid.py to GLE_critical_ca_advanced.py

- Better reflects that this is an advanced critical Ca finder
- Uses a three-phase approach, not true continuation for both branches
- Maintains consistency with naming convention

VatsalSy committed:
Add compatibility wrapper for GLE_continuation_hybrid.py

- Creates a deprecation warning for users of the old filename
- Redirects to the new GLE_critical_ca_advanced.py
- Ensures backward compatibility during transition

VatsalSy committed:
Add modular utility files in src-local directory

- ca_continuation_utils.py: Critical Ca finding utilities
- gle_utils.py: Core GLE solver functions
- solution_types.py: Data structures for solutions
- solver_utils.py: Additional solver utilities

These modules extract common functionality from GLE_solver.py
for better code organization and reusability

VatsalSy committed:
Update GLE_solver.py: Refactor and improve critical Ca finder

- Extract common functions to src-local modules
- Fix x0 marker in plots (now shows point instead of vertical line)
- Rename output files: pyGLE_profiles.png, pyGLE_Ca_cr.png, pyGLE_solution.csv
- Update function names to reflect root-finding approach (not continuation)
- Fix critical Ca finder to use refined solution for final solve
- Change command-line args: --lambda-slip to --lambda_slip

VatsalSy committed:
Update GLE_critical_ca_advanced.py: Command-line argument changes

- Change --lambda-slip to --lambda_slip for consistency
- Update after rename from GLE_continuation_hybrid.py
- Maintains three-phase advanced critical Ca finding approach

VatsalSy committed:
Update README.md: Documentation improvements and consistency

- Clarify GLE_solver.py uses root-finding (NOT continuation) for Ca_cr
- Update all command examples: --lambda-slip to --lambda_slip, --mu-r to --mu_r
- Update output filenames: pyGLE_profiles.png, pyGLE_Ca_cr.png, pyGLE_solution.csv
- Add note about GLE_critical_ca_advanced.py as advanced Ca finder
- Improve clarity on differences between solvers

VatsalSy committed:
Update CLAUDE.md: Project documentation and command updates

- Document GLE_critical_ca_advanced.py (renamed from GLE_continuation_hybrid.py)
- Update command examples: --lambda-slip to --lambda_slip, --mu-r to --mu_r
- Clarify that GLE_solver.py finds Ca_cr via root-finding, not continuation
- Add notes about the three-phase approach in the advanced finder
- Update known issues section for the renamed file

VatsalSy committed:
Refactor GLE_solver.py for improved readability and add time tracking

Major improvements:
- Modularized run_solver_and_plot() into smaller, focused functions
- Added comprehensive time tracking for all stages:
  * Direct solve attempt
  * Critical Ca search (Stage 1, 1.5, and 2)
  * Final solve at Ca_cr
  * Plotting and I/O operations
- Enhanced ca_continuation_utils.py with detailed timing for each stage
- Improved code readability with clear step-by-step organization
- Added print_timing_summary() for detailed performance analysis
- Better separation of concerns and function documentation

The timing information helps identify performance bottlenecks and
provides insights into which stages take the most computational time.

VatsalSy committed:
Rename critical Ca solvers for consistency and refactor GLE_solver

- Rename GLE_critical_ca_advanced.py to GLE_criticalCa_advanced.py
- Add new GLE_criticalCa.py for standard critical Ca finding
- Refactor GLE_solver.py for improved simplicity and readability
- Update utility modules for consistency

VatsalSy committed:
Improve output protocol consistency in GLE_criticalCa.py

- Clarify plot titles to show only actual Ca used (remove requested Ca reference)
- Add clear messaging when requested Ca exceeds Ca_critical
- Return Ca_actual from solver to properly report which value was used
- Change parameter display from "Ca:" to "Ca_requested:" for clarity
- Ensure output protocol matches GLE_criticalCa_advanced.py behavior

VatsalSy committed:
Update documentation and enhance critical Ca solver with adaptive mesh refinement

- Update CLAUDE.md to accurately reflect current codebase structure and available solvers
- Clarify distinction between GLE_solver.py (simple solver), GLE_criticalCa.py (standard critical Ca finder), and GLE_criticalCa_advanced.py (advanced with adaptive mesh)
- Add comprehensive documentation for GLE_continuation_hybrid.py pseudo-arclength continuation
- Enhance GLE_criticalCa_advanced.py with improved adaptive mesh refinement and better error handling
- Remove deprecated Phase 1/2/3 documentation to reflect simplified single-phase approach
- Update command examples and usage guidelines

The advanced critical Ca finder now provides more robust convergence with adaptive mesh density
adjustments based on solution properties, while maintaining the proven hybrid IQI/Newton/bisection
refinement approach from the standard solver.

VatsalSy committed:
output added

VatsalSy committed:
added a naivee continuation method

VatsalSy committed:
remove pkkl

VatsalSy committed:
sort of works.. not sure through

amanb1901 committed:
The sign in F(theta, R) denominator was wrong is Chan et al., 2013). Corrected it.

amanb1901 committed:
The denominator in F(theta, R) had a sign error in Chan et al., 2013). Fixed now.

amanb1901 committed:
F(\theta, R) had a sign error in Chan et al., 2013. Fixed now.

amanb1901 committed:
Correct form of GLE

VatsalSy committed:
Refactor continuation methods and remove deprecated simple implementation

- Remove GLE_continuation_simple.py (replaced by v1 implementation)
- Add GLE_continuation_v1.py with improved continuation algorithm
- Update GLE_continuation_v0.py with enhanced functionality

VatsalSy committed:
Update documentation to reflect continuation method changes

- Update README.md with new v1 continuation implementation details
- Document improved algorithm features and usage

VatsalSy committed:
Update solver output profiles with latest results

- Refresh pyGLE_solver-profiles.csv with improved data formatting
- Update corresponding visualization in pyGLE_solver-profiles.png
